# Telecom Churn Case Study Hackathon

# Problem Statement

In the telecom industry, customers are able to choose from multiple service providers and actively switch from one operator to another. In this highly competitive market, the telecommunications industry experiences an average of 15-25% annual churn rate. Given the fact that it costs 5-10 times more to acquire a new customer than to retain an existing one, customer retention has now become even more important than customer acquisition.

For many incumbent operators, retaining high profitable customers is the number one business
goal. To reduce customer churn, telecom companies need to predict which customers are at high risk of churn. In this project, you will analyze customer-level data of a leading telecom firm, build predictive models to identify customers at high risk of churn.

In this competition, your goal is to build a machine learning model that is able to predict churning customers based on the features provided for their usage.

# Business Objectives

Predict churning customers for a Telecom company based on temporal behaviour

# IMPORT LIBRARIES AND DATASETS

In [1]:
# import package
import numpy as np                                                 # Implemennts milti-dimensional array and matrices
import pandas as pd                                                # For data manipulation and analysis
import matplotlib.pyplot as plt                                    # Plotting library for Python programming language and it's numerical mathematics extension NumPy
import seaborn as sns                                              # Provides a high level interface for drawing attractive and informative statistical graphics

pd.set_option('display.max_columns', None)
%matplotlib inline
sns.set()

In [2]:
import warnings
warnings.filterwarnings('ignore')

## Data Loading and Description

- **train.csv** - Contains 172 columns. The primary key that represents each customer is id. The target variable that you need to predict is churn_probability which contains a value of 0 or 1. This data is what you are going to use for EDA, cleaning, feature engineering, model building, model evaluation, model selection, and finally model training.
- **test.csv** - Contains 171 columns, doesn't contain the target variable churn_probability. You will apply all the necessary preprocessing steps to get this data into the right format and then use the model trained using the train.csv file to make predictions with this. This is unseen data! Only the competition hosts know the actual values of the target feature for this data and therefore your submissions will be evaluated on how well your model does with this dataset.
- **sample.csv** - This contains the format in which you need to submit the solutions to Kaggle. The id column in this dataset exactly the same as the id column in test.csv. You will make your predictions on the test.csv data and store them in a submission file that has the same format as this file. Check the Overview>Evaluation tab for mode details.
- **Data Dictionary** - This contains the definitions for the various acronyms that you will need to understand each variable. For example, the variable total_og_mou_7, contains the acronyms total, og, mou, and 7, which can be translated as the total outgoing minutes of voice calls made by the user in month of July.

To gain insights from data we must look into each aspect of it very carefully. We will start with observing few rows and columns of data both from the starting and from the end.

Let us check the basic information of the dataset. The very basic information to know is the dimension of the dataset – rows and columns – that’s what we find out with the method shape.

In [3]:
df = pd.read_csv("/kaggle/input/telecom-churn-case-study-hackathon-c42/train.csv")

In [4]:
df.sample(5)

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
69126  69126        109             0.0             0.0             0.0   
5654    5654        109             0.0             0.0             0.0   
27473  27473        109             0.0             0.0             0.0   
26926  26926        109             0.0             0.0             0.0   
58245  58245        109             0.0             0.0             0.0   

      last_date_of_month_6 last_date_of_month_7 last_date_of_month_8   arpu_6  \
69126            6/30/2014            7/31/2014            8/31/2014  184.244   
5654             6/30/2014            7/31/2014            8/31/2014  701.137   
27473            6/30/2014            7/31/2014            8/31/2014  182.690   
26926            6/30/2014            7/31/2014            8/31/2014  276.230   
58245            6/30/2014            7/31/2014            8/31/2014  258.295   

        arpu_7    arpu_8  onnet_mou_6  onnet_mou_7  onnet_mou_8  offnet_mou_6  \
69126  209.710   357.042        34.01        23.46        22.34        156.39   
5654   571.352  1234.312       439.36       355.33       663.79        259.53   
27473   99.070    97.841        68.43        26.86        25.68        109.83   
26926  186.499   181.450        48.54        21.36         0.00         57.66   
58245  276.889   109.940       112.84        68.74        44.04         81.26   

       offnet_mou_7  offnet_mou_8  roam_ic_mou_6  roam_ic_mou_7  \
69126         94.34        247.11            0.0            0.0   
5654         220.66        523.31            0.0            0.0   
27473         87.68         36.69            0.0            0.0   
26926          0.30          0.28            0.0            0.0   
58245         87.31         44.06            0.0            0.0   

       roam_ic_mou_8  roam_og_mou_6  roam_og_mou_7  roam_og_mou_8  \
69126            0.0            0.0            0.0            0.0   
5654             0.0            0.0            0.0            0.0   
27473            0.0            0.0            0.0            0.0   
26926            0.0            0.0            0.0            0.0   
58245            0.0            0.0            0.0            0.0   

       loc_og_t2t_mou_6  loc_og_t2t_mou_7  loc_og_t2t_mou_8  loc_og_t2m_mou_6  \
69126             33.09              3.83             20.08            156.39   
5654             439.36            355.33            663.79            252.36   
27473             68.43             26.86             25.68            102.79   
26926             48.54             21.36              0.00             53.26   
58245            110.81             67.61             44.04             81.26   

       loc_og_t2m_mou_7  loc_og_t2m_mou_8  loc_og_t2f_mou_6  loc_og_t2f_mou_7  \
69126             93.69            228.54              0.00              0.00   
5654             206.93            488.58              7.16              6.94   
27473             76.69             31.81              2.01             10.26   
26926              0.00              0.28              0.00              0.00   
58245             87.11             44.06              0.00              0.20   

       loc_og_t2f_mou_8  loc_og_t2c_mou_6  loc_og_t2c_mou_7  loc_og_t2c_mou_8  \
69126              0.00               0.0               0.0              0.00   
5654              34.73               0.0               0.0              0.00   
27473              2.01               0.0               0.0              2.86   
26926              0.00               0.0               0.3              0.00   
58245              0.00               0.0               0.0              0.00   

       loc_og_mou_6  loc_og_mou_7  loc_og_mou_8  std_og_t2t_mou_6  \
69126        189.49         97.53        248.63              0.91   
5654         698.89        569.21       1187.11              0.00   
27473        173.24        113.83         59.51              0.00   
26926        101.81         21

# Data Understanding

In [5]:
df.shape

(69999, 172)

The dataset has __69999 rows and 172 columns.__

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Columns: 172 entries, id to churn_probability
dtypes: float64(135), int64(28), object(9)
memory usage: 91.9+ MB


In [7]:
# Get info from both data set to understand data in depths 
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69999 entries, 0 to 69998
Data columns (total 172 columns):
 #    Column                    Dtype  
---   ------                    -----  
 0    id                        int64  
 1    circle_id                 int64  
 2    loc_og_t2o_mou            float64
 3    std_og_t2o_mou            float64
 4    loc_ic_t2o_mou            float64
 5    last_date_of_month_6      object 
 6    last_date_of_month_7      object 
 7    last_date_of_month_8      object 
 8    arpu_6                    float64
 9    arpu_7                    float64
 10   arpu_8                    float64
 11   onnet_mou_6               float64
 12   onnet_mou_7               float64
 13   onnet_mou_8               float64
 14   offnet_mou_6              float64
 15   offnet_mou_7              float64
 16   offnet_mou_8              float64
 17   roam_ic_mou_6             float64
 18   roam_ic_mou_7             float64
 19   roam_ic_mou_8             float64
 20   roam

We have following 9 Columns as object type
- last_date_of_month_6
- last_date_of_month_7
- last_date_of_month_8
- date_of_last_rech_6
- date_of_last_rech_7
- date_of_last_rech_8
- date_of_last_rech_data_6
- date_of_last_rech_data_7
- date_of_last_rech_data_8

In [8]:
df.describe()

id  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count  69999.000000    69999.0         69297.0         69297.0   
mean   34999.000000      109.0             0.0             0.0   
std    20207.115084        0.0             0.0             0.0   
min        0.000000      109.0             0.0             0.0   
25%    17499.500000      109.0             0.0             0.0   
50%    34999.000000      109.0             0.0             0.0   
75%    52498.500000      109.0             0.0             0.0   
max    69998.000000      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8   onnet_mou_6  \
count         69297.0  69999.000000  69999.000000  69999.000000  67231.000000   
mean              0.0    283.134365    278.185912    278.858826    133.153275   
std               0.0    334.213918    344.366927    351.924315    299.963093   
min               0.0  -2258.709000  -1289.715000   -945.808000      0.000000   
25%               0.0     93.581000     86.714000     84.095000      7.410000   
50%               0.0    197.484000    191.588000    192.234000     34.110000   
75%               0.0    370.791000    365.369500    369.909000    119.390000   
max               0.0  27731.088000  35145.834000  33543.624000   7376.710000   

        onnet_mou_7   onnet_mou_8  offnet_mou_6  offnet_mou_7  offnet_mou_8  \
count  67312.000000  66296.000000  67231.000000  67312.000000  66296.000000   
mean     133.894438    132.978257    198.874771    197.153383    196.543577   
std      311.277193    311.896596    316.818355    322.482226    324.089234   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        6.675000      6.410000     34.860000     32.240000     31.575000   
50%       32.280000     32.100000     96.480000     91.885000     91.800000   
75%      115.837500    115.060000    232.990000    227.630000    229.345000   
max     8157.780000  10752.560000   8362.360000   7043.980000  14007.340000   

       roam_ic_mou_6  roam_ic_mou_7  roam_ic_mou_8  roam_og_mou_6  \
count   67231.000000   67312.000000   66296.000000   67231.000000   
mean        9.765435       7.014568       7.004892      14.186457   
std        57.374429      55.960985      53.408135      73.469261   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max      2850.980000    4155.830000    4169.810000    3775.110000   

       roam_og_mou_7  roam_og_mou_8  loc_og_t2t_mou_6  loc_og_t2t_mou_7  \
count   67312.000000   66296.000000      67231.000000      67312.000000   
mean        9.842191       9.771783         46.904854         46.166503   
std        58.511894      64.618388        150.971758        154.739002   
min         0.000000       0.000000          0.000000          0.000000   
25%         0.000000       0.000000          1.660000          1.650000   
50%         0.000000       0.000000         11.910000         11.580000   
75%         0.000000       0.000000         40.740000         39.760000   
max      2812.040000    5337.040000       6431.330000       7400.660000   

       loc_og_t2t_mou_8  loc_og_t2m_mou_6  loc_og_t2m_mou_7  loc_og_t2m_mou_8  \
count      66296.000000      67231.000000      67312.000000      66296.000000   
mean          45.686109         93.238231         90.799240         91.121447   
std          153.716880        162.046699        153.852597        152.997805   
min            0.000000          0.000000          0.000000          0.000000   
25%            1.610000          9.920000         10.090000          9.830000   
50%           11.740000         41.030000         40.170000         40.350000   
75%           39.895000        110.430000        107.540000        109.245000   
max        10752.560000       4696

#### Utils functions

In [9]:
# Checking more meta information about dataframe i.e. how many non-null, uniquie value.

def metadataInfo(dataframe) : 
    return pd.DataFrame({
                'Datatype' : dataframe.dtypes.astype(str), 
                'Non_Null_Count': dataframe.count(axis = 0).astype(int), 
                'Null_Count': dataframe.isnull().sum().astype(int), 
                'Null_Percentage': round(dataframe.isnull().sum()/len(dataframe) * 100 , 2), 
                'Unique_Values_Count': dataframe.nunique().astype(int) 
                 }).sort_values(by='Null_Percentage', ascending=False)

In [10]:
new_df = metadataInfo(df)
print(new_df.to_string())

                         Datatype  Non_Null_Count  Null_Count  Null_Percentage  Unique_Values_Count
arpu_3g_6                 float64           17568       52431            74.90                 5506
count_rech_2g_6           float64           17568       52431            74.90                   30
night_pck_user_6          float64           17568       52431            74.90                    2
arpu_2g_6                 float64           17568       52431            74.90                 5389
date_of_last_rech_data_6   object           17568       52431            74.90                   30
total_rech_data_6         float64           17568       52431            74.90                   36
av_rech_amt_data_6        float64           17568       52431            74.90                  792
max_rech_data_6           float64           17568       52431            74.90                   47
count_rech_3g_6           float64           17568       52431            74.90                   23


- The dataset comprises of 69999 observations of 172 columns. Above is a table showing names of all the columns and their description.
- There are few Null/NA value in dataset.
- Some columns that are categorical in nature, but in integer/float type.
- Also few columns have object type as date value which we can use to create a derive data from that

# Data Engineering 

### Data Cleanning

# Feature Engineering and Selection

# Exploratory Data Analysis 

# Model Validation & Evaluation

# Model Predict

# Conclusion 